Wrap unqualified imports

oneness opened the pull request:
Addresses #610 

Introduces flex break to allow vector of Document to break or unbreak
based on the set limit (currently 80).  

At its core, this is just about adding line breaks where ast of
UnqualifiedImports are more than the current limit. I think I followed
the intended design of decoupling parser specific types from
formatting where the main abstraction is the Document enum
type (Please correct me if I am wrong). This is my first Rust program
and could be done better if we only had to traverse the ast just
once (had to iter on it by cloning). I'd appreciate if you provide any
feedback on how to improve it.

Other than that, I think it is pretty generic enough to be
reused. Again I could be totally wrong and any feedback are welcome.

lpil commented on src/format/tests.rs:
```diff
@@ -25,6 +25,8 @@ fn module_test() {
     assert_format!("import one/two/three as free\n");
     assert_format!("import one/two/three.{thunk} as free\n");
     assert_format!("import one/two/three.{thunk as funky} as free\n");
+    assert_format!("import my/cool/module.{\nOne, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Eleven, Twelve, \nThirteen, Fifteen, Sixteen\n}\n");
```
Could we have some indentation after the break? Currently it is formatted like this:

```rust
import my/cool/module.{
One, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Eleven, Twelve, 
Thirteen, Fifteen, Sixteen
}
```

When I think it should be formatted like this:

```rust
import my/cool/module.{
  One, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Eleven, Twelve, 
  Thirteen, Fifteen, Sixteen,
}
```

There is also some trailing whitespace on the second line.

oneness replied:
The fix is checked in.

oneness commented:
Your feedback other than the trailing whitespace bug is addressed and have a look to let me know if that is the right direction taken.  Thanks.

lpil commented on src/format.rs:
```diff
@@ -207,13 +207,23 @@ impl<'a> Formatter<'a> {
+                                        break_(&e.clone().name, &e.clone().name).append(
+                                            if let Some(name) = &e.as_name {
+                                                format!(" as {}", name).to_doc()
+                                            } else {
```
I think this will render a break between the name of the field being imported and the `as` as `break_` can insert a newline after it.  Could we not have the previous logic and intersperse a `delim(",")`?

oneness replied:
I refactored the code a bit to make it more readable.  Import as is not a `Break Document` so new line will not be inserted.  Please let me know what you think.  Will address any concern/feedback you have.  Thanks.

lpil replied:
Thanks :)

oneness commented:
> Looking good!
> 
> I've found an example that get formatted incorrectly (it changes the name of the imported type)
> 
> ```rust
> import gleam/result.{
>   Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
>   Abcde, End
> }
> ```
 
Thanks. Nice example. It caught couple of bugs:
1.  Should remove last trailing space char not just any chars. 
2. The fit logic should have checked for `broken` length since it is what is being padded. 

So your example will be formatted like this as it should be: 
```
import gleam/result.{
  Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa, Abcde,
  End
}
```
Since `Aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa, Abcde,` is 78 chars long, which fits within the 80 chars limit.  

I fixed both bug and added above to the test cases in the latest commit.

lpil commented on src/format.rs:
```diff
@@ -207,19 +207,21 @@ impl<'a> Formatter<'a> {
+                                unqualified
+                                    .iter()
+                                    .map(|e| {
+                                        break_(&e.clone().name, &e.clone().name)
+                                            .append(import_as(&e.clone().as_name))
```
We can reduce the amount of memory allocation like so:
```suggestion
                                        e.name
                                            .to_doc()
                                            .append(break_("", ""))
                                            .append(import_as(&e.as_name))
```

lpil commented on src/pretty.rs:
```diff
@@ -209,6 +248,18 @@ fn fmt(b: &mut String, limit: isize, mut width: isize, mut docs: Vector<(isize,
+                    docs.push_front((
+                        indent,
+                        Mode::Flex,
+                        line().nest(2).append((*doc).clone()).append(line()),
```
The document should not have hard coded indents in it, instead we should apply the nested when constructing the algebra.

With hard coded nesting this won't work for Erlang, which uses 4 space indentation, or for plain text, which doesn't indent at all.

lpil commented on src/pretty.rs:
```diff
@@ -182,6 +209,18 @@ fn fmt(b: &mut String, limit: isize, mut width: isize, mut docs: Vector<(isize,
+                        if limit - width < broken.len() as isize {
+                            remove_ws_end(b, 1);
+                            b.push_str("\n");
+                            b.push_str(" ".repeat(2).as_str());
```
Same indentation situation here, it should respect the current indent level rather than hard coding 2.

lpil commented on src/pretty.rs:
```diff
@@ -182,6 +209,18 @@ fn fmt(b: &mut String, limit: isize, mut width: isize, mut docs: Vector<(isize,
                     }
+                    Mode::Flex => {
+                        if limit - width < broken.len() as isize {
+                            remove_ws_end(b, 1);
```
This doesn't seem right to me. Why do we have trailing whitespace? The Document algebra shouldn't know anything about the content it contains, the formatting rules should be entirely in the structure of the algebra and not conditional on whether it contains content. I think the best approach here is to not construct a Document that renders trailing whitesapce, as it currently does on master.
